### DISCUSSION TITLE
extending variants

### DISCUSSION THREAD

**Discussion by violabg** [chakra-ui/chakra-ui#2974] [Q&A]
is it possible to extend a variant?

For example I want to create a variant for a button and would like to be very similar to the outline verison, is it possible to spread the outline variant to my new variant and override only what I need?

```js
variants: {
    secondary: (props: any) => {
      // const variants = props.theme.components.Button.variants;
      // console.log("variants.outline", variants.outline);
      return {
        // ...variants.outline,
        color: "white",
        border: "1px solid white",
        _hover: {
          backgroundColor: "brand.500",
          borderColor: "brand.500"
        }
      };
    }
  }
```

**Comment by chasinhues** [comment]
Good question, looks like it is possible. You can import `theme` and spread the variant you want to extend over your new variant! 

### Demo
https://codesandbox.io/s/extend-variants-demo-0nhtc?file=/src/theme/index.ts

**Reply by sh-sam** [reply]
Hi @chasinhues - I would like to change the default `_hover` and `_active` _shades_ for button variants dynamically for each `colorScheme` color.

Currently if I have a set of buttons like this:

`<Button variant="outline" colorScheme="blue">a button</Button>`
`<Button variant="outline" colorScheme="red">a button</Button>`
`<Button variant="outline" colorScheme="yellow">a button</Button>`

The default shade for `_hover` will be `blue.50`, `red.50`, `yellow.50`- respectively.

Is it possible to use props to dynamically and uniformly change the hover colors to `blue.25`, `red.25`, and `yellow.25`?

The idea is that I could do something like this, based on the codesandbox link you shared:

```
variants: {
	outline: (props) => ({
		...theme.components.Button.variants.outline(props),
		borderWidth: 2,
		_hover: {
			backgroundColor: `${props.currentColor}.25`,
		},
	}),
},
```

Is something like this possible?

**Comment by violabg** [comment]
Thank you,
I had tried that before asking, but it doesn't really work. if you look at your example, it does not really extend the outline variable.
here if the code:

https://codesandbox.io/s/extend-variants-demo-forked-pexll?file=/src/theme/index.ts

**Reply by chasinhues** [reply]
Ah sorry, my bad, I was fooled by the demo 🤣  I don't know of any other way in that case -- you might need to bite the bullet and build a variant from scratch with the styles you want.

I've created a feature request for this though, as I've encountered a situation before where I've wanted to extend variants as well.

https://github.com/chakra-ui/chakra-ui/issues/3118

**Comment by dodas** [comment]
Hey guys.
Spreading the default variant is possible, but Button's variants are defined as functions, not objects, so you will need to call it first:
```ts
variants: {
  secondary: (props) => ({
    ...theme.components.Button.variants.outline(props),
    color: "brand.500",
    _hover: {
      color: "white",
      backgroundColor: "brand.500",
      borderColor: "brand.500",
    }
  })
}
```
Working example: https://codesandbox.io/s/extend-variants-demo-forked-ylish?file=/src/App.tsx

**Reply by cmco** [reply]
Hi! Do you know how to also do this for colors? Since they're just objects, I tried accessing them with the spread operator normally but still couldn't access them.

**Reply by stefee** [reply]
@dodas @pom421 - This solution no longer works with strict TypeScript since `theme.components` now has the type `Record<string, any>` this results in ts-eslint error

```
Unsafe member access .variants on an `any` value
```

Sad times!

**Reply by stefee** [reply]
You can use type assertions as a a workaround e.g.

```ts
const defaultTableTheme = defaultTheme.components.Table as ComponentStyleConfig;
const defaultVariant = defaultTableTheme.variants?.simple as PartsStyleFunction;
```

**Comment by pom421** [comment]
Extending the solution above, with a multi part component. 

An Input is composed of field, addon and element. Let's style the field part with a different border, based on the outline variant. 

```js
import defaultTheme from "@chakra-ui/theme"
import { StyleFunctionProps } from "@chakra-ui/theme-tools"
```

```js
  components: {
    Input: {
      variants: {
        "blue-outline": (props: StyleFunctionProps) => ({
          ...defaultTheme.components.Input.variants.outline(props),
          field: {
            ...defaultTheme.components.Input.variants.outline(props).field,
            border: "1px solid",
            borderColor: "blue.100",
          },
        }),
      },
    },
  },
```

**Reply by elboletaire** [reply]
Are you using `definePartsStyle` by any chance? Using it I cannot workaround the type errors, since does not look to be expecting a function at all 😕
